File: matlab_prototype/stft_kkmw.m

% STFT including zero padding (optional)
% x: multichannel signal (sign_len,nr_mics) [length in samples]
% M: frame length (samples)
% N: fft length (samples), zeropadding for N > M
% R: hop-size (samples)

function [X, f_vec, t] = stft_kkmw(x,M,N,R,fs)

% set defaults
if nargin < 5
    error('Input arguments missing');
end

% pre-allocation
%num_frames = ceil(size(x,1)/R);
num_frames = floor((size(x,1)-M)/R)+1;
N_out = N/2+1;
nr_mics = size(x,2);

xt = zeros(M,nr_mics,'double');
f_vec = (0:N/2) * fs/N;
t = (0:num_frames-1) * R/fs;
X = zeros(N_out,num_frames,nr_mics,'double');

% transform window
%win = repmat(sin(linspace(0,pi,M)).',[1 nr_mics]);
win = repmat(hamming(M),[1 nr_mics]);

% STFT
for kk = 0:num_frames-1
    %xt = [ xt(R+1:end,:); x(kk*R+1:(kk+1)*R,:) ];
    xt = x((kk*R+1):(kk*R+M), :);
    Xt = fft(xt .* win, N);
    X(:,kk+1,1:nr_mics) = Xt(1:N_out,:);
end

File: python_prototype/ADMM/stft_kkmw.py
from numpy.ma import floor
import numpy as np

def stft_kkmw(x,M,N,R,fs):

    num_frames = (floor((len(x)-M)/R+1))
    N_out= N/2+1
    nr_mics = len(x[0])

    xt = np.zeros((M, nr_mics))
    f_vec = np.arange(N/2+1) * fs/N
    t = np.arange(num_frames)*R/fs
    X=np.zeros((int(N_out),int(num_frames),int(nr_mics)),dtype=np.complex_)

    win = np.transpose(np.matlib.repmat(np.hamming(M), nr_mics, 1))

    for kk in range (0, int(num_frames)):
        xt=x[kk*R:kk*R+M]
        xt=xt*win
        Xt = np.fft.fft(xt, N,0)
        X[:,kk, 0:nr_mics] = Xt[0:int(N_out)]

    return X